code/modules/holodeck/computer.dm

```diff
@@ -110,63 +120,193 @@
+	if (program == map_id)
+		return
+
+	if (!is_operational)//load_program is called once with a timer (in toggle_power) we dont want this to load anything if its off
+		map_id = offline_program
+		force = TRUE
+
+	if (current_cd > world.time && !force)
+		say("ERROR. Recalibrating projection apparatus.")
+		return
+
+	if (add_delay)
+		current_cd = world.time + HOLODECK_CD
+		if(damaged)
+			current_cd += HOLODECK_DMG_CD
+	active = (map_id != offline_program)
+	use_power = active + IDLE_POWER_USE
+	program = map_id
+
+	//clear the items from the previous program
+	if (spawned)
+		for (var/_item in spawned)
+			var/obj/holo_item = _item
+			derez(holo_item)
+	if (effects)
+		for (var/_effect in effects)
+			var/obj/effect/holodeck_effect/holo_effect = _effect
+			effects -= holo_effect
+			holo_effect.deactivate(src)
+
+	template = SSmapping.holodeck_templates[map_id]
+	for (var/_turf in linked)
+		var/turf/holo_turf = _turf
+		if (istype(holo_turf, /turf/closed))
+			for (var/_baseturf in holo_turf.baseturfs)
+				if (ispath(_baseturf, /turf/open/floor/holofloor))
+					holo_turf.baseturfs -= _baseturf
+					holo_turf.baseturfs += /turf/open/floor/holofloor/plating
+
+	template.load(bottom_left)//this is what actually loads the holodeck simulation into the map
+
+	spawned = template.created_atoms
+
+	nerf(!(obj_flags & EMAGGED))
+	finish_spawn()
+
+/obj/machinery/computer/holodeck/proc/finish_spawn()//this is used for holodeck effects (like spawners). otherwise they dont do shit
+	for (var/_atom in spawned)
+		var/atom/atoms = _atom
+
+		if (isturf(atoms))
+			var/turf/holo_turf = atoms
+			spawned -= holo_turf
+
+		atoms.flags_1 |= HOLOGRAM_1
+
+		if (istype(atoms, /obj/effect/holodeck_effect/))//this is what makes holoeffects work
+			var/obj/effect/holodeck_effect/holo_effect = atoms
+			effects += holo_effect
+			spawned -= holo_effect
+			var/atom/active_effect = holo_effect.activate(src)
+			if(istype(active_effect) || islist(active_effect))
+				spawned += active_effect // holocarp are not forever
+
+		if (isobj(atoms))
+			var/obj/holo_object = atoms
+			holo_object.resistance_flags = LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF
+
+			if (ismachinery(holo_object))
+				var/obj/machinery/machines = holo_object
+				machines.flags_1 |= NODECONSTRUCT_1
+				machines.power_change()
+
+				if(istype(machines, /obj/machinery/button))
+					var/obj/machinery/button/buttons = machines
+					buttons.setup_device()
+
+			if (isstructure(holo_object))
+				holo_object.flags_1 |= NODECONSTRUCT_1
+
+///this qdels holoitems that should no longer exist for whatever reason
+/obj/machinery/computer/holodeck/proc/derez(obj/object, silent = TRUE, forced = FALSE)
+	if(!object)
+		return
+
+	spawned -= object
+	var/turf/target_turf = get_turf(object)
+	for(var/atom/movable/object_contents in object) // these should be derezed if they were generated
+		object_contents.forceMove(target_turf)
```

Comment:
Why is this move necessary here ?

Reply:
leftover from original holodeck code that i didnt get rid of, ill take it out in the next update (assuming it doesnt ruin anything but i dont see why it would)

Reply:
found out why its necessary, if a non holodeck item is inside of a holodeck container (like the morgue trays), they can get qdel'd with the container without this proc. so far it only seems to work like that with the morgue tray, I assume that closets do it themselves on deletion.